### PR TITLE
net: ip: ipv4: Add fragmented packet support

### LINKED ISSUE #49783: net: ipv4: packet fragmentation support
**Is your enhancement proposal related to a problem? Please describe.**
Adding support for fragmented packet reassembly (and optionally, transmission) for IPv4 networking. This allows for packets > 1500 bytes or smaller but fragmented packets if links have smaller MTUs.

**Describe the solution you'd like**
For IPv4 fragmentation to be supported, similar to how IPv6 fragmentation is

**Describe alternatives you've considered**
None

### PR DESCRIPTION
**Pull request by thedjnK**
Adds support for incoming and outgoing IPv4 fragmented packet support, allowing a single packet that is too large to be sent to be split up and sent successfully.

Fixes #49783

Tested using `smp_svr` sample on `native_posix_64` and a Cortex M board with ethernet, test passes with `native_posix_64` on local PC.

### CODE REVIEW

File: subsys/net/ip/ipv4_fragment.c [hunk 1]
```diff
@@ -0,0 +1,647 @@
...
+#include "tcp_internal.h"
+#include "ipv4.h"
+#include "route.h"
+#include "net_stats.h"
+
+/* Timeout for various buffer allocations in this file. */
+#define NET_BUF_TIMEOUT K_MSEC(100)
+
+/* How long to wait for a buffer */
+#define FRAG_BUF_WAIT K_MSEC(10)
```
Comment (rlubos) [inline subsys/net/ip/ipv4_fragment.c:31]: It seems that this define is not used anywhere?

**Review by rlubos — COMMENTED** [review]
Thank you for the effort, the implementation seems pretty neat. Do you plan to add some basic tests for the IPv4 fragmentation? Currently, the code wouldn't even be built by the CI, not even mentioning possible future regressions.

File: include/zephyr/net/net_pkt.h [hunk 2]
```diff
@@ -234,6 +234,11 @@ struct net_pkt {
 	uint16_t vlan_tci;
 #endif /* CONFIG_NET_VLAN */
 
+#if defined(CONFIG_NET_IPV4) && defined(CONFIG_NET_IPV4_FRAGMENT)
+	uint16_t ipv4_fragment_flags;	/* Fragment offset and M (More Fragment) flag */
+	uint16_t ipv4_fragment_id;	/* Fragment ID */
```
[the remark this answers is not in the document]
Comment (nordicjm) [reply include/zephyr/net/net_pkt.h:239]: Will have a play.

Not sure on what to do for testing at present, the one idea I do have is testing it using mcumgr to send a very long echo and checking it is received and sent back, will need to look at the network tests to see if this is possible and how they work though.

[the remark this answers is not in the document]
Comment (rlubos) [reply include/zephyr/net/net_pkt.h:239]: There is an existing test suite for IPv6 fragmentation:
https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/net/ipv6_fragment

It's not a rule of thumb, but typically network tests either create a dummy interface to capture raw IP packets or loopback for local communication.

**Review by tbursztyka — COMMENTED** [review]
would be nice to get a commit adding test for this feature (as it has been done for ipv6 fragment support).

File: subsys/net/ip/ipv4_fragment.c [hunk 3]
```diff
@@ -0,0 +1,647 @@
...
+		struct net_pkt *pkt = reass->pkt[i];
+		unsigned int offset;
+		int payload_len;
+
+		if (!pkt) {
+			break;
+		}
+
+		offset = net_pkt_ipv4_fragment_offset(pkt);
+
```
[the remark this answers is not in the document]
Comment (thedjnK) [reply subsys/net/ip/ipv4_fragment.c:269]: The line gap is how it is in other files, e.g. ipv6_fragment.c

File: include/zephyr/net/net_pkt.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (thedjnK) [reply include/zephyr/net/net_pkt.h]: Fixed

File: include/zephyr/net/net_pkt.h [hunk 4]
```diff
@@ -234,6 +234,11 @@ struct net_pkt {
 	uint16_t vlan_tci;
 #endif /* CONFIG_NET_VLAN */
 
+#if defined(CONFIG_NET_IPV4) && defined(CONFIG_NET_IPV4_FRAGMENT)
```
[the remark this answers is not in the document]
Comment (thedjnK) [reply include/zephyr/net/net_pkt.h:238]: Fixed

File: subsys/net/ip/net_core.c [hunk 5]
```diff
@@ -71,11 +72,12 @@ static inline enum net_verdict process_data(struct net_pkt *pkt,
...
 	}
 
-	/* If the packet is routed back to us when we have reassembled
-	 * an IPv6 packet, then do not pass it to L2 as the packet does
-	 * not have link layer headers in it.
+	/* If the packet is routed back to us when we have reassembled an IPv4 or IPv6 packet,
+	 * then do not pass it to L2 as the packet does not have link layer headers in it.
 	 */
-	if (IS_ENABLED(CONFIG_NET_IPV6_FRAGMENT) && net_pkt_ipv6_fragment_start(pkt)) {
+	if (IS_ENABLED(CONFIG_NET_IPV4_FRAGMENT) && net_pkt_ipv4_fragment_more(pkt)) {
```
[the remark this answers is not in the document]
Comment (thedjnK) [reply subsys/net/ip/net_core.c]: Fixed

File: subsys/net/ip/net_core.c [hunk 6]
```diff
@@ -448,6 +450,10 @@ static inline void l3_init(void)
 	net_icmpv6_init();
 	net_ipv6_init();
 
+	if (IS_ENABLED(CONFIG_NET_IPV4_FRAGMENT)) {
```
[the remark this answers is not in the document]
Comment (thedjnK) [reply subsys/net/ip/net_core.c:453]: Fixed

**Comment by thedjnK** [comment]
Will have to make some time to create tests for this

File: include/zephyr/net/net_pkt.h [hunk 7]
```diff
@@ -234,6 +234,22 @@ struct net_pkt {
 	uint16_t vlan_tci;
 #endif /* CONFIG_NET_VLAN */
 
+	union {
+#if defined(CONFIG_NET_IPV4) && defined(CONFIG_NET_IPV4_FRAGMENT)
```
Comment (jukkar) [inline include/zephyr/net/net_pkt.h:238]: Do we need the first part of the check here, wouldn't `#if defined(CONFIG_NET_IPV4_FRAGMENT)` be enough as the fragment symbol is not defined if IPv4 is not enabled.
Same comment for IPv6 check.

**Comment by jukkar** [comment]
Minor nit about the commit subject, it would be enough to set it to
`net: ipv4: Add fragmented packet support `

**Review by krish2718 — DISMISSED** [review]

**Comment by nordicjm** [comment]
Tests are about 33% complete thus far, will aim to get them finished and submitted to PR within the next week.

File: include/zephyr/net/net_pkt.h (hunk 7 above)
Comment (thedjnK) [reply include/zephyr/net/net_pkt.h]: Fixed

File: subsys/net/ip/net_if.c [hunk 8]
```diff
@@ -484,6 +485,12 @@ enum net_verdict net_if_send_data(struct net_if *iface, struct net_pkt *pkt)
 		verdict = net_ipv6_prepare_for_send(pkt);
 	}
 
+#if defined(CONFIG_NET_IPV4) && defined(CONFIG_NET_IPV4_FRAGMENT)
```
[the remark this answers is not in the document]
Comment (thedjnK) [reply subsys/net/ip/net_if.c]: Fixed

**Comment by thedjnK** [comment]
@jukkar @tbursztyka @rlubos Updated with corrections and test/release notes added. There is one outstanding issue with the test which is to check the ID, and since this is generated randomly I'm not sure I can get it from the test to verify it, other than that it's all ready for a review.

File: tests/net/ipv4_fragment/src/main.c [hunk 9]
```diff
@@ -0,0 +1,886 @@
...
+		net_pkt_cursor_init(recv_pkt);
+		net_pkt_skip(recv_pkt, 12);
+		net_pkt_write(recv_pkt, buf_ip_dst, sizeof(buf_ip_dst));
+		net_pkt_write(recv_pkt, buf_ip_src, sizeof(buf_ip_src));
+
+		/* Get offset to check if this is the first packet */
+		net_pkt_cursor_init(recv_pkt);
+		net_pkt_skip(recv_pkt, 6);
+		net_pkt_read_be16(recv_pkt, &offset);
+		offset &= 0b1111111111111;
```
Comment (jukkar) [inline tests/net/ipv4_fragment/src/main.c:318]: Hmm, this does not look like a standard C but is only available as a gcc extension. I suggest that normal hex value is used here instead.

**Review by jukkar — COMMENTED** [review]
LGTM, only comment about the bit constant that imho should be changed to hex constant.

File: tests/net/ipv4_fragment/src/main.c (hunk 9 above)
Comment (nordicjm) [reply tests/net/ipv4_fragment/src/main.c]: Fixed

**Review by jukkar — APPROVED** [review]

**Review by rlubos — APPROVED** [review]
LGTM

**Pull request merged**